src/coreclr/jit/gentree.h

```diff
@@ -350,6 +351,341 @@ struct Statement;
+    GTF_ICON_FTN_ADDR           = 0xD0000000, // GT_CNS_INT -- constant is a function address
+    GTF_ICON_CIDMID_HDL         = 0xE0000000, // GT_CNS_INT -- constant is a class ID or a module ID
+    GTF_ICON_BBC_PTR            = 0xF0000000, // GT_CNS_INT -- constant is a basic block count pointer
+
+    GTF_ICON_FIELD_OFF          = 0x08000000, // GT_CNS_INT -- constant is a field offset
+    GTF_ICON_SIMD_COUNT         = 0x04000000, // GT_CNS_INT -- constant is Vector<T>.Count
+
+    GTF_ICON_INITCLASS          = 0x02000000, // GT_CNS_INT -- Constant is used to access a static that requires preceding
+                                              //               class/static init helper.  In some cases, the constant is
+                                              //               the address of the static field itself, and in other cases
+                                              //               there's an extra layer of indirection and it is the address
+                                              //               of the cell that the runtime will fill in with the address
+                                              //               of the static field; in both of those cases, the constant
+                                              //               is what gets flagged.
+
+    GTF_BLK_VOLATILE            = GTF_IND_VOLATILE,  // GT_ASG, GT_STORE_BLK, GT_STORE_OBJ, GT_STORE_DYNBLK -- is a volatile block operation
+    GTF_BLK_UNALIGNED           = GTF_IND_UNALIGNED, // GT_ASG, GT_STORE_BLK, GT_STORE_OBJ, GT_STORE_DYNBLK -- is an unaligned block operation
+
+    GTF_OVERFLOW                = 0x10000000, // Supported for: GT_ADD, GT_SUB, GT_MUL and GT_CAST.
+                                              // Requires an overflow check. Use gtOverflow(Ex)() to check this flag.
+
+    GTF_DIV_BY_CNS_OPT          = 0x80000000, // GT_DIV -- Uses the division by constant optimization to compute this division
+
+    GTF_ARR_BOUND_INBND         = 0x80000000, // GT_ARR_BOUNDS_CHECK -- have proved this check is always in-bounds
+
+    GTF_ARRLEN_ARR_IDX          = 0x80000000, // GT_ARR_LENGTH -- Length which feeds into an array index expression
+    GTF_ARRLEN_NONFAULTING      = 0x20000000, // GT_ARR_LENGTH  -- An array length operation that cannot fault. Same as GT_IND_NONFAULTING.
+
+    GTF_SIMDASHW_OP             = 0x80000000, // GT_HWINTRINSIC -- Indicates that the structHandle should be gotten from gtGetStructHandleForSIMD
+                                              //                   rather than from gtGetStructHandleForHWSIMD.
+
+    // Flag used by assertion prop to indicate that a type is a TYP_LONG
+#ifdef TARGET_64BIT
+    GTF_ASSERTION_PROP_LONG     = 0x00000001,
+#endif // TARGET_64BIT
+};
+
+inline constexpr GenTreeFlags operator ~(GenTreeFlags a)
+{
+    return (GenTreeFlags)(~(unsigned int)a);
+}
+
+inline constexpr GenTreeFlags operator |(GenTreeFlags a, GenTreeFlags b)
```

Comment:
Are you planning to add more enums like that in the future? If so maybe declare these methods as `#define ENUM_METHODS` and then call it for `GenTreeDebugFlags`, `GenTreeFlags`, `GenTreeDebugFlags`? 

Or, if we require C++20 then we can do it with templates with restrictions:
```
template<typename T> requires is_enum<T>::value inline constexpr T operator ~(T a)

{
    return (T)(~(std::underlying_type_t<T>)a);
}
```
but if I remember correctly it is pretty fresh so maybe a significant change in min. supported version.

Reply:
maybe you can add `MAKE_*` there as well.

Reply:
> Are you planning to add more enums like that in the future?

I don't think there are any more to add. Yes, there is some obvious duplication in the code between them, but then again, the code is very simple. I'd lean against adding macros to define things; I'd prefer to avoid macros as much as possible. Your C++20 case looks interesting, but we basically can't use STL / std:: namespace. Maybe in the future?

Reply:
MAKE_BBFLAG makes sense for the block flags because, at least right now, each one is separate, independent, and non-overlapping. I think it doesn't make sense for the GTF_ flags because there are overlapping cases for different classes of objects, there are ranges where a flag isn't actually just a bit, but a number in part of a range (namely, the GTF_ICON_ values), etc.